File: new/darwin/tab.m
// 12 april 2015
#import "uipriv_darwin.h"

// TODO
// - verify margins against extra space around the tab

struct tab {
	uiTab t;
	NSTabView *tabview;
	NSMutableArray *pages;
};

static void destroy(void *data)
{
	struct tab *t = (struct tab *) data;

	[t->pages enumerateObjectsUsingBlock:^(id obj, NSUInteger index, BOOL *stop) {
		NSValue *v = (NSValue *) obj;
		uiContainer *p;

		// TODO this is definitely wrong but
		p = (uiContainer *) [v pointerValue];
		uiControlDestroy(uiControl(p));
	}];
	[t->pages release];
	uiFree(t);
}

// the default new control implementation uses -sizeToFit, which we don't have with NSTabView
// fortunately, we do have -minimumSize
static void preferredSize(uiControl *c, uiSizing *d, intmax_t *width, intmax_t *height)
{
	struct tab *t = (struct tab *) c;
	NSSize s;

	s = [t->tabview minimumSize];
	*width = (intmax_t) (s.width);
	*height = (intmax_t) (s.height);
}

static void tabAppendPage(uiTab *tt, const char *name, uiControl *child)
{
	struct tab *t = (struct tab *) tt;
	uiContainer *page;
	NSTabViewItem *i;

	page = newBin();
	binSetMainConotrol(page, child);
	[t->pages addObject:[NSValue valueWithPointer:page]];

	i = [[NSTabViewItem alloc] initWithIdentifier:nil];
	[i setLabel:toNSString(name)];
	[i setView:((NSView *) uiContainerHandle(content))];
	[t->tabview addTabViewItem:i];
}

static void tabDeletePage(uiTab *tt, uintmax_t n)
{
	struct tab *t = (struct tab *) tt;
	NSValue *v;
	uiContainer *p;
	NSTabViewItem *i;

	v = (NSValue *) [t->pages objectAtIndex:n];
	p = (uiContainer *) [v pointerValue];
	[t->pages removeObjectAtIndex:n];

	// make sure the children of the tab aren't destroyed
	binSetMainControl(p, NULL);

	// TODO negotiate lifetimes better
	i = [t->tabview tabViewItemAtIndex:n];
	[t->tabview removeTabViewItem:i];
}

uiTab *uiNewTab(void)
{
	struct tab *t;

	t = uiNew(struct tab);

	uiDarwinNewControl(uiControl(t), [NSTabView class], NO, NO, destroy, t);

	t->tabview = (NSTabView *) uiControlHandle(uiControl(t));

	// also good for NSTabView (same selector and everything)
	setStandardControlFont((NSControl *) (t->tabview));

	t->pages = [NSMutableArray new];

	uiControl(t)->PreferredSize = preferredSize;

	uiTab(t)->AppendPage = tabAppendPage;
	uiTab(t)->DeletePage = tabDeletePage;

	return uiTab(t);
}

File: new/unix/tab.c
// 12 april 2015
#include "uipriv_unix.h"

struct tab {
	uiTab t;
	GtkWidget *widget;
	GtkContainer *container;
	GtkNotebook *notebook;
	// TODO switch to GArray or GPtrArray
	uiContainer **pages;
	uintmax_t len;
	uintmax_t cap;
};

static void onDestroy(void *data)
{
	struct tab *t = (struct tab *) data;
	uintmax_t i;

	// first hide ourselves to avoid flicker
	gtk_widget_hide(t->widget);
	// the pages do not have a libui parent, so we can simply destroy them
	// TODO verify if this is sufficient
	for (i = 0; i < t->len; i++)
		uiControlDestroy(uiControl(t->pages[i]));
	// then free ourselves
	uiFree(t->pages);
	uiFree(t);
}

#define tabCapGrow 32

static void tabAppendPage(uiTab *tt, const char *name, uiControl *child)
{
	struct tab *t = (struct tab *) tt;
	uiContainer *page;
	GtkWidget *pageWidget;

	if (t->len >= t->cap) {
		t->cap += tabCapGrow;
		t->pages = (uiContainer **) uiRealloc(t->pages, t->cap * sizeof (uiContainer *), "uiContainer *[]");
	}

	page = newBin();
	binSetMainControl(page, child);
	// and add it as a tab page
	binSetParent(page, (uintptr_t) (t->container));
	pageWidget = GTK_WIDGET(uiControlHandle(uiControl(page)));
	gtk_notebook_set_tab_label_text(t->notebook, pageWidget, name);

	t->pages[t->len] = page;
	t->len++;
}

static void tabDeletePage(uiTab *tt, uintmax_t n)
{
	struct tab *t = (struct tab *) tt;
	uiContainer *page;
	uintmax_t i;

	page = t->pages[n];
	for (i = n; i < t->len - 1; i++)
		t->pages[i] = t->pages[i + 1];
	t->pages[i] = NULL;
	t->len--;

	// make sure the page's control isn't destroyed
	binSetMainControl(page, NULL);

	// now destroy the page
	// this will also remove the tab
	// why? simple: both gtk_notebook_remove_tab() and gtk_widget_destroy() call gtk_container_remove()
	// TODO verify this
	uiControlDestroy(uiControl(page));
}

uiTab *uiNewTab(void)
{
	struct tab *t;

	t = uiNew(struct tab);

	uiUnixNewControl(uiControl(t), GTK_TYPE_NOTEBOOK,
		FALSE, FALSE, onDestroy, t,
		NULL);

	t->widget = GTK_WIDGET(uiControlHandle(uiControl(t)));
	t->container = GTK_CONTAINER(t->widget);
	t->notebook = GTK_NOTEBOOK(t->widget);

	uiTab(t)->AppendPage = tabAppendPage;
	uiTab(t)->DeletePage = tabDeletePage;

	return uiTab(t);
}
